packages/@glimmer/component/addon/-private/component.ts

```diff
@@ -2,12 +2,89 @@ import { DEBUG } from '@glimmer/env';
+export let ARGS_SET: ArgsSetMap;
 
 if (DEBUG) {
-  ARGS_SET = new WeakMap();
+  ARGS_SET = new WeakMap() as ArgsSetMap;
 }
 
+// --- Type utilities for component signatures --- //
+// Type-only "symbol" to use with `EmptyObject` below, so that it is *not*
+// equivalent to an empty interface.
+declare const Empty: unique symbol;
+
+/**
+ * This provides us a way to have a "fallback" which represents an empty object,
+ * without the downsides of how TS treats `{}`. Specifically: this will
+ * correctly leverage "excess property checking" so that, given a component
+ * which has no named args, if someone invokes it with any named args, they will
+ * get a type error.
+ *
+ * @internal This is exported so declaration emit works (if it were not emitted,
+ *   declarations which fall back to it would not work). It is *not* intended for
+ *   public usage, and the specific mechanics it uses may change at any time.
+ *   The location of this export *is* part of the public API, because moving it
+ *   will break existing declarations, but is not legal for end users to import
+ *   themselves, so ***DO NOT RELY ON IT***.
+ */
+export type EmptyObject = { [Empty]?: true };
+
+type GetOrElse<Obj, K extends PropertyKey, Fallback> = Obj extends { [Key in K]: infer U }
+  ? U
+  : Fallback;
+
+/** Given a signature `S`, get back the `Args` type. */
+type ArgsFor<S> = S extends { Args: infer Args }
+  ? Args extends { Named?: object; Positional?: unknown[] } // Are they longhand already?
+    ? {
+        Named: GetOrElse<S['Args'], 'Named', EmptyObject>;
+        Positional: GetOrElse<S['Args'], 'Positional', []>;
+      }
+    : { Named: S['Args']; Positional: [] }
+  : { Named: EmptyObject; Positional: [] };
+
+/**
+ * Given any allowed shorthand form of a signature, desugars it to its full
+ * expanded type.
+ *
+ * @internal This is only exported so we can avoid duplicating it in
+ *   [Glint](https://github.com/typed-ember/glint) or other such tooling. It is
+ *   *not* intended for public usage, and the specific mechanics it uses may
+ *   change at any time. Although the signature produced by is part of Glimmer's
+ *   public API the existence and mechanics of this specific symbol are *not*,
+ *   so ***DO NOT RELY ON IT***.
+ */
+export type ExpandSignature<T> = {
+  Element: GetOrElse<T, 'Element', null>;
+  Args: keyof T extends 'Args' | 'Element' | 'Blocks' // Is this a `Signature`?
+    ? ArgsFor<T> // Then use `Signature` args
+    : { Named: T; Positional: [] }; // Otherwise fall back to classic `Args`.
+  Blocks: T extends { Blocks: infer Blocks }
+    ? {
+        [Block in keyof Blocks]: Blocks[Block] extends unknown[]
+          ? { Positional: Blocks[Block] }
+          : Blocks[Block];
+      }
+    : EmptyObject;
+};
+
+/**
+ * @internal we use this type for convenience internally; inference means users
+ *   should not normally need to name it
+ */
+export type Args<S> = ExpandSignature<S>['Args']['Named'];
```

Comment:
Ah, interesting. That was not the intent, and we actually explicitly said we *weren’t* going to do that—the intent of that part of the RFC was to show how we would do it *internally* in the type signature. Accordingly, we should start by shipping without it: when people *need* to name args for their own constructor, they can always just reference the signature directly. We can then see what happens in the community over time: Do people end up writing that helper a lot in practice, or requesting it? If so, we can follow up and supply it later.

Reply:
I think the helper is particularly useful internally because of all the different forms a signature might take: with or without `Args`, with or without the `Named`/`Positional` expansion, etc.

For a user, though, they have their concrete signature right there and know what they wrote.

```ts
interface GreetingSig {
 Args: { message: string };
}

class Greeting extends Component<GreetingSig> {
  constructor(owner: Owner, args: GreetingSig['Args']) {
    // ...
  }
}
```

Or I've seen folks with Glint define `GreetingArgs` and then `GreetingSig` in terms of that (which is how we discovered the whole "can't use an `interface` for a type param that `extends Record<...>` thing):

```ts
interface GreetingArgs {
  message: string;
}

interface GreetingSig {
  Args: GreetingArgs;
}

class Greeting extends Component<GreetingSig> {
  constructor(owner: Owner, args: GreetingArgs) {
    // ...
  }
}
```

Something like this would also work, I think:

```ts
class Greeting extends Component<GreetingSig> {
  constructor(...params: ConstructorParameters<Component<GreetingSig>>) {
    super(...params);
    // ...
  }
}
```

Reply:
I know that we said we _weren't_ going to make this public, but then I went back and looked over the RFC and for some reason read it as saying we _were_ going to expose this helper. Re-reading it again, though, it's clear that this is internal for the type of `@glimmer/component` and not for end users. 😆

I agree that end users can easily get the args type either by making it a separate interface or pulling it out of the signature they have defined. (BTW, `ConstructorParameters` doesn't work here because it needs a class type so you'd have to do `ConstructorParameters<typeof Component>`, but then lose the generic).

Reply:
Ah right. In 4.7 `ConstructorParameters<typeof Component<Sig>>` will work, though!

https://www.typescriptlang.org/play?ts=4.7.0-dev.20220330#code/MYGwhgzhAECyCeBhcUA8AVAfNA3gKGmmAHsA7CAFwCcBXYC4qgCjCoHMIAuaAQSqrDwMmAJS4ChImQjEQAUwB0IYmxbsIIgNwSAvnj14K8AA5zo6aAF5oiadToMqABVZgAtnIpyqEVEdPEAGZwSCi+lFQAlqRsmJiaQA

Reply:
Oh, nice!
/me checks to see what other goodies are coming in 4.7